### PR TITLE
[Data] Don't convert variable-length byte data to Numpy array

### LINKED ISSUE #35586: [CI] `linux://doc:source/data/doc_code/loading_data` is failing/flaky on master.
- 44ef068bd1edbe3e54818c22ccd12a7dd50937f1 FAILED [Buildkite :book: Doc tests and examples (excluding Ray AIR examples)](https://buildkite.com/ray-project/oss-ci-build-branch/builds/4031#018840f7-40fa-4b80-a02f-bd7936e75e4f)
- 323d9d5f2c4c503b24afb20831bcd0370b505e88 FAILED [Buildkite :book: Doc tests and examples (excluding Ray AIR examples)](https://buildkite.com/ray-project/oss-ci-build-branch/builds/4030#01883f8a-ce4d-475d-bd3f-28e8eee933cf)
- 0a1f435534487da0970d99a7c81122869f15e8bc FAILED [Buildkite :book: Doc tests and examples (excluding Ray AIR examples)](https://buildkite.com/ray-project/oss-ci-build-branch/builds/4029#01883f83-9463-4409-a694-acf4445651da)
- 946459a63eb82fbc9c9e60beb867d6fd3e3aa849 FAILED [Buildkite :book: Doc tests and examples (excluding Ray AIR examples)](https://buildkite.com/ray-project/oss-ci-build-branch/builds/4028#01883eb0-f9c1-434b-aef9-58b293f80b82)
- 63ebee22f726117b7682fdcc38da8fb7a7ecdbd3 FAILED [Buildkite :book: Doc tests and examples (excluding Ray AIR examples)](https://buildkite.com/ray-project/oss-ci-build-branch/builds/4027#01883d1c-aea0-44c7-9859-dda1fee6c569)
- 3275be3d9e6017bf93fa33df4e215877d156a6ea FAILED [Buildkite :book: Doc tests and examples (excluding Ray AIR examples)](https://buildkite.com/ray-project/oss-ci-build-branch/builds/4025#01883af9-35ad-4856-b727-280876f353bf)
- c0f04f24e6f66be8a13589d5b61b383b6ff0c89f FAILED [Buildkite :book: Doc tests and examples (excluding Ray AIR examples)](https://buildkite.com/ray-project/oss-ci-build-branch/builds/4023#01883748-8aed-455e-a489-f39e3878bdc2)
- b3d97e844b89b2fedbf4d59c33bed1e3150868a0 FAILED [Buildkite :book: Doc tests and examples (excluding Ray AIR examples)](https://buildkite.com/ray-project/oss-ci-build-branch/builds/4022#01883645-82ae-4902-921f-a82c045d929e)
- 06d25a76f5d14870185ddfec5db16f3e6412fa4b FAILED [Buildkite :book: Doc tests and examples (excluding Ray AIR examples)](https://buildkite.com/ray-project/oss-ci-build-branch/builds/4013#01883500-8e92-493b-8ec6-ef544e8d5152)
- 17f23191a4af1d88d3122dfdf28b49e2086716cc FAILED [Buildkite :book: Doc tests and examples (excluding Ray AIR examples)](https://buildkite.com/ray-project/oss-ci-build-branch/builds/4012#01883422-9e29-46cf-b3d2-994f14919b16)
- f170d13a92e2b99f8c44ca927214b051394dcee0 FAILED [Buildkite :book: Doc tests and examples (excluding Ray AIR examples)](https://buildkite.com/ray-project/oss-ci-build-branch/builds/4011#0188327b-8d77-46b9-b8b2-30b6a66fe5a5)
- ce7764b635a443507630f8840c906e68f10a1ebc FAILED [Buildkite :book: Doc tests and examples (excluding Ray AIR examples)](https://buildkite.com/ray-project/oss-ci-build-branch/builds/4008#01883190-f55f-43f6-9739-d229cfe32108)
- 43a20c130ddebec8d9a495fd0aaab69e5fee71ac FAILED [Buildkite :book: Doc tests and examples (excluding Ray AIR examples)](https://buildkite.com/ray-project/oss-ci-build-branch/builds/4007#0188318a-a346-4b7d-a835-11c5ff244c6c)
- 35305e45614a8ff47345d72ddd87317a30d26a1b FAILED [Buildkite :book: Doc tests and examples (excluding Ray AIR examples)](https://buildkite.com/ray-project/oss-ci-build-branch/builds/4005#01883164-9f85-4ed9-8da5-ed2a41b8b7ad)
- 35305e45614a8ff47345d72ddd87317a30d26a1b FAILED [Buildkite :book: Doc tests and examples (excluding Ray AIR examples)](https://buildkite.com/ray-project/oss-ci-build-branch/builds/4005#01883597-ef25-4a06-ad3f-cf664e6f11e9)
- ec94bd422dd2c07327d620797a214e150a4c00a9 FAILED [Buildkite :book: Doc tests and examples (excluding Ray AIR examples)](https://buildkite.com/ray-project/oss-ci-build-branch/builds/4004#01883153-74bb-4db8-aac9-20a32a425e1d)
- 72205c6bfb87cde9df730862aa3b085466b75584 FAILED [Buildkite :book: Doc tests and examples (excluding Ray AIR examples)](https://buildkite.com/ray-project/oss-ci-build-branch/builds/4003#01883149-fe4e-4d68-bbb2-24fe2c0f771b)
- de3df47447f1db547f1844298593f5243a4a077a FAILED [Buildkite :book: Doc tests and examples (excluding Ray AIR examples)](https://buildkite.com/ray-project/oss-ci-build-branch/builds/4001#0188310a-bf0e-4b7f-9e3d-8b9bfffefacb)
- 092d25b99e1a15e36bf0bee2d52f4277e4b1e016 FAILED [Buildkite :book: Doc tests and examples (excluding Ray AIR examples)](https://buildkite.com/ray-project/oss-ci-build-branch/builds/4000#01883108-4933-40eb-a960-bc0e6a66f8a8)
- f1b7eb668ec3bf1ff1d4fe3187179f778f1482d0 FAILED [Buildkite :book: Doc tests and examples (excluding Ray AIR examples)](https://buildkite.com/ray-project/oss-ci-build-branch/builds/3999#01883104-a5e8-4447-95d6-52237de3af74)

....
Generated from flaky test tracker. Please do not edit the signature in this section.
DataCaseName-linux://doc:source/data/doc_code/loading_data-END
....

### PR DESCRIPTION
**Pull request by amogkam**
Closes #35586 
See https://github.com/ray-project/ray/issues/35586#issuecomment-1558148261

Numpy treats variable length byte data as zero-terminated bytes. So if there are zero bytes encoded into the bytestring itself, those will be discarded.

Instead, per recommendation in https://github.com/apache/arrow/issues/26470, it seems that variable length bytes should be treated as python objects.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by ArturNiederfahrenhorst** [comment]
Please also create a cherry pick PR for 2.5.0 when this is done :)

**Review by raulchen — APPROVED** [review]
LGTM. Also add a unit test in `test_numpy_support.py`?

**Comment by amogkam** [comment]
good point, added!

File: python/ray/data/_internal/numpy_support.py [hunk 1]
```diff
@@ -72,16 +72,25 @@ def convert_udf_returns_to_numpy(udf_return_col: Any) -> Any:
...
                 if isinstance(e, np.ndarray):
                     shapes.add((e.dtype, e.shape))
+                elif isinstance(e, bytes):
+                    # Don't convert variable length binary data to Numpy arrays as it
+                    # treats zero encoding as termination by default.
+                    # Per recommendation from
+                    # https://github.com/apache/arrow/issues/26470,
+                    # we use object dtype.
+                    # https://github.com/ray-project/ray/issues/35586#issuecomment-1558148261
+                    dtype = object
```
Comment (ericl) [inline python/ray/data/_internal/numpy_support.py:86]: Would be simpler to set has_object=True. All create ragged ndarray does is create an object array under the hood.

**Pull request merged**